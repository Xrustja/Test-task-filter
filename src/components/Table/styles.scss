.table {
  &__title {
    font-size: 28px;
    color: #2777b7;
    text-align: center;
    margin-bottom: 30px;
  }

  .filter__block {
    display: flex;
    justify-content: flex-start;
    align-items: center;
    font-size: 18px;

    .filter__input {
      border: 2px solid #12264f;
      border-radius: 3px;
      height: 46px;
      width: 200px;
      margin-right: 5px;
      box-sizing: border-box;
      padding: 5px 10px;
      font-size: 18px;

      &:focus-visible {
        box-shadow: 0 4px 32px rgba(39, 119, 183, 0.4);
        outline: none;
      }
    }

    .filter__btn {
      outline: none;
      background-color: #12264f;
      color: white;
      border: none;
      border-radius: 3px;
      cursor: pointer;
      min-height: 46px;
      padding: 5px 20px;
      font-size: 20px;
    }
  }

  .table {
    font-size: 16px;
    color: #2777b7;

    &__container {
      border: 2px solid #2777b7;
      margin-top: 25px;
    }

    &__row {
      border-bottom: 2px solid #2777b7;
      display: flex;
      flex-wrap: nowrap;
      width: 100%;

      &:last-child {
        border-bottom: none;
      }

      &.heading {
        background-color: lightblue;
      }
    }
  }

  & .loading {
    align-self: center;
    display: flex;
    justify-content: center;

    .fa {
      font-size: 40px;
      margin: 20px 0;
    }
  }

  .btn-loadMore {
    outline: none;
    background-color: #12264f;
    color: white;
    border: none;
    border-radius: 3px;
    cursor: pointer;
    min-height: 46px;
    padding: 3px 20px;
    font-size: 18px;
    margin: 10px auto;
    display: block;
  }
}
