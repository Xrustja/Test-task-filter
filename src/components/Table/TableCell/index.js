import React, {useEffect, useState} from "react";
import './styles.scss';

const TableCell = ({data, heading, filter, activeSortType}) => {
    const [active, setActive] = useState('');

    function activateSort(type) {
        filter(data, type)
    }

    useEffect(() => {
        if (activeSortType && activeSortType.code === data.code) {
            setActive(activeSortType.type);
        } else {
            setActive('');
        }
    }, [activeSortType])

    return (
        heading
            ?
            <div className="cell">
                <p className="cell__name">{data.name}</p>
                <div className="cell__filters">
                    <i className={`${(active === 'asc')
                        ? 'fas fa-caret-up active'
                        : 'fas fa-caret-up'}`}
                       onClick={() => activateSort('asc')}/>
                    <i className={`${(active === 'desc')
                        ? 'fas fa-caret-down active'
                        : 'fas fa-caret-down'}`}
                       onClick={() => activateSort('desc')}
                    />
                </div>
            </div>
            :
            <div className="table__row">
                <div className="cell">{data.name}</div>
                <div className="cell">{data.birth_year}</div>
                <div className="cell">{data.height}</div>
                <div className="cell">{data.mass}</div>
            </div>

    )
};
export default TableCell;
