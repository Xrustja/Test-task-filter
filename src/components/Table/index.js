import React, {useEffect, useState} from "react";
import APIRequests from "../../services/APIRequests";
import getCharacters from "../../services/getCharacters";
import './styles.scss';
import TableCell from "./TableCell";

const {get} = APIRequests;

const Table = () => {
        const [characters, setCharacters] = useState([]);
        const [filteredCharacters, setFilteredCharacters] = useState([]);
        const [page, setPage] = useState('');
        const [name, setName] = useState('');
        const [activeSortType, setActiveSortByType] = useState({code: null, type: null});
        const [loading, setLoading] = useState(false);
        const [hideBtn, setHideBtn] = useState(false);
        let tableHeadingNames = [{name: 'Title', code: 'name'}, {name: 'Birth Year', code: 'birth_year'}, {
            name: 'Height', code: 'height'
        }, {name: 'Mass', code: 'mass'}]


        const loadMore = () => {
            setLoading(true);
            get(page)
                .then(({results: nextCharacters, next}) => {
                    setPage(next);
                    let list = [...characters, ...nextCharacters];
                    setCharacters(list);
                    setFilteredCharacters(list);
                    if (name) {
                        setFilteredCharacters(filter(list))
                    }
                    next === null ? setHideBtn(true) : setHideBtn(false);
                    setLoading(false);
                })
        };
        useEffect(() => {
            let cleanupFunction = false;
            getCharacters()
                .then(({results: characters, next}) => {
                    if (!cleanupFunction) {
                        setCharacters(characters || []);
                        setFilteredCharacters(characters || [])
                        setPage(next || null)
                    }
                })
            return () => {
                cleanupFunction = true
            }
        }, []);


        function filterByName(e) {
            e.preventDefault();
            if (name !== "") {
                setFilteredCharacters(filter(characters))
            }
        }

        function filter(list) {
            return list.filter(character => character.name.toLowerCase().includes(name.toLowerCase()))
        }

        function checkIsNaN(a) {
            return isNaN(a) ? a : +a
        }

        function sortByParams(column, type, list) {
            setActiveSortByType({code: column.code, type: type})
            setFilteredCharacters((list || filteredCharacters).sort(function (a, b) {
                a = checkIsNaN(a[column.code])
                b = checkIsNaN(b[column.code])

                if (a > b) {
                    return type === "asc" ? 1 : -1;
                }
                if (a < b) {
                    return type === "asc" ? -1 : 1;
                }
                return 0;
            }))
        }


        return (
            <div className='table'>
                <h2 className="table__title"> StarWars characters table info</h2>
                <form className="filter__block" id="comment-form"
                      onSubmit={e => {
                          filterByName(e)
                      }}>
                    <input type="text" placeholder='Type name here'
                           className="filter__input"
                           value={name}
                           onChange={(e => setName(e.target.value))}
                           name="name"/>
                    <button type='submit' className="filter__btn">Filter</button>
                </form>
                <div className="table__container">
                    <div className="table__row heading">
                        {tableHeadingNames.map(cell => {
                            return (
                                <TableCell data={cell} key={cell.name} heading={true} filter={sortByParams}
                                           activeSortType={activeSortType}/>
                            )
                        })}

                    </div>
                    {characters.length === 0
                        ?

                        <div className="loading">
                            <i className="fa fa-spinner fa-spin"/>
                        </div>

                        :
                        <>
                            {filteredCharacters.map(character => {
                                return (
                                    <TableCell data={character} key={character.name} heading={false}/>
                                )
                            })}
                        </>
                    }
                </div>
                <button className={`btn-loadMore ${hideBtn ? 'hidden' : ''}`} onClick={loadMore}>
                    {loading ? <i className="fa fa-spinner fa-spin"/> : ''}Load more
                </button>
            </div>
        )
    }
;
export default Table;
