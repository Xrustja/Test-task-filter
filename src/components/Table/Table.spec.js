import '@testing-library/jest-dom'
import axios from "axios";
import renderer, {act} from "react-test-renderer";
import Table from "./index";

jest.mock('axios');

const mockedData = {
    "results": [
        {
            "name": "Luke Skywalker",
            "birth_year": "123",
            "height": "172",
            "mass": "77"
        },
        {
            "name": "C-3PO",
            "birth_year": "123",
            "height": "167",
            "mass": "75"
        }]
};

beforeEach(() => {
    axios.get.mockResolvedValue({data: mockedData})
});

test('component renders correctly', async () => {
    let table;
    await act(async () => {
        table = renderer.create(<Table/>);
    })
    expect(table.toJSON()).toMatchSnapshot();
})
