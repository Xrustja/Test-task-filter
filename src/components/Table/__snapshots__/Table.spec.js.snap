// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component renders correctly 1`] = `
<div
  className="table"
>
  <h2
    className="table__title"
  >
     StarWars characters table info
  </h2>
  <form
    className="filter__block"
    id="comment-form"
    onSubmit={[Function]}
  >
    <input
      className="filter__input"
      name="name"
      onChange={[Function]}
      placeholder="Type name here"
      type="text"
      value=""
    />
    <button
      className="filter__btn"
      type="submit"
    >
      Filter
    </button>
  </form>
  <div
    className="table__container"
  >
    <div
      className="table__row heading"
    >
      <div
        className="cell"
      >
        <p
          className="cell__name"
        >
          Title
        </p>
        <div
          className="cell__filters"
        >
          <i
            className="fas fa-caret-up"
            onClick={[Function]}
          />
          <i
            className="fas fa-caret-down"
            onClick={[Function]}
          />
        </div>
      </div>
      <div
        className="cell"
      >
        <p
          className="cell__name"
        >
          Birth Year
        </p>
        <div
          className="cell__filters"
        >
          <i
            className="fas fa-caret-up"
            onClick={[Function]}
          />
          <i
            className="fas fa-caret-down"
            onClick={[Function]}
          />
        </div>
      </div>
      <div
        className="cell"
      >
        <p
          className="cell__name"
        >
          Height
        </p>
        <div
          className="cell__filters"
        >
          <i
            className="fas fa-caret-up"
            onClick={[Function]}
          />
          <i
            className="fas fa-caret-down"
            onClick={[Function]}
          />
        </div>
      </div>
      <div
        className="cell"
      >
        <p
          className="cell__name"
        >
          Mass
        </p>
        <div
          className="cell__filters"
        >
          <i
            className="fas fa-caret-up"
            onClick={[Function]}
          />
          <i
            className="fas fa-caret-down"
            onClick={[Function]}
          />
        </div>
      </div>
    </div>
    <div
      className="table__row"
    >
      <div
        className="cell"
      >
        Luke Skywalker
      </div>
      <div
        className="cell"
      >
        123
      </div>
      <div
        className="cell"
      >
        172
      </div>
      <div
        className="cell"
      >
        77
      </div>
    </div>
    <div
      className="table__row"
    >
      <div
        className="cell"
      >
        C-3PO
      </div>
      <div
        className="cell"
      >
        123
      </div>
      <div
        className="cell"
      >
        167
      </div>
      <div
        className="cell"
      >
        75
      </div>
    </div>
    <button
      className="btn btn-loadMore "
      onClick={[Function]}
    >
      
      Load more
    </button>
  </div>
</div>
`;
